contracts/modules/andromeda-shunting/src/contract.rs

```diff
@@ -0,0 +1,188 @@
+            ado_type: "shunting".to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            operators: None,
+            kernel_address: msg.kernel_address,
+            owner: msg.owner,
+        },
+    )?;
+
+    Ok(inst_resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let _contract = ADOContract::default();
+    let ctx = ExecuteContext::new(deps, info, env);
+
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        }
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+pub fn handle_execute(ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateExpressions { expressions } => {
+            execute_update_expression(ctx, expressions)
+        }
+        _ => ADOContract::default().execute(ctx, msg),
+    }
+}
+
+fn execute_update_expression(
+    ctx: ExecuteContext,
+    expressions: Vec<String>,
+) -> Result<Response, ContractError> {
+    let ExecuteContext { deps, info, .. } = ctx;
+    nonpayable(&info)?;
+
+    ensure!(
+        ADOContract::default().is_owner_or_operator(deps.storage, info.sender.as_str())?,
+        ContractError::Unauthorized {}
+    );
+
+    EXPRESSIONS.save(deps.storage, &expressions)?;
+
+    Ok(Response::new().add_attributes(vec![attr("action", "update_expression")]))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<Binary, ContractError> {
+    match msg {
+        QueryMsg::Evaluate { params } => encode_binary(&handle_eval_expression(deps, params)?),
+        _ => ADOContract::default().query(deps, env, msg),
+    }
+}
+
+fn handle_eval_expression(
+    deps: Deps,
+    params: Vec<EvaluateParam>,
+) -> Result<ShuntingResponse, ContractError> {
+    let expressions = EXPRESSIONS.load(deps.storage)?;
+    let mut results: Vec<f64> = Vec::new();
+    let mut result: f64 = 0.0;
+
+    // replace parameters that require other shunting
+    let params = parse_params(deps, params)?;
+
+    for (index, expression) in expressions.iter().enumerate() {
+        let mut parsed_expression = expression.to_string();
+
+        // replace x0, x1 ... with actual params in expression
+        for (ndx, param) in params.iter().enumerate() {
+            let placeholder = format!("x{}", ndx);
+            parsed_expression = parsed_expression.replace(&placeholder, param);
+        }
+
+        // replace y0, y1 ... with calculation results
+        for (i, sub_result) in results.iter().enumerate().take(index) {
+            let placeholder = format!("y{}", i);
+            parsed_expression = parsed_expression.replace(&placeholder, &sub_result.to_string());
+        }
+
+        result = eval(&parsed_expression).unwrap();
```

Comment:
Replace .unwrap() with error mapping to tell the user which expression errored

Reply:
`eval` function is now returning `ContractError` with details.